File: src/domain/Wallet.ts
import { InsufficientFundsError } from './errors/InsufficientFundsError.js';
import { InvalidAmountError } from './errors/InvalidAmountError.js';

export interface WalletProps {
  id: string;
  currency: string;
  balance: number;
}

export class Wallet {
  readonly id: string;
  readonly currency: string;
  private _balance: number;

  constructor({ id, currency, balance }: WalletProps) {
    this.id = id;
    this.currency = currency;
    this._balance = balance;
  }

  get balance(): number {
    return this._balance;
  }

  debit(amount: number): void {
    if (amount <= 0) {
      throw new InvalidAmountError(amount);
    }
    if (this._balance < amount) {
      throw new InsufficientFundsError(this._balance, amount);
    }
    this._balance = Number((this._balance - amount).toFixed(2));
  }
}

File: internal/domain/wallet.go
package domain

import (
	"errors"
)

var (
	ErrCurrencyMismatchWallet = errors.New("wallet currency mismatch")
	ErrInsufficientFunds      = errors.New("insufficient funds")
)

// Wallet represents a customer wallet with a single currency balance.
type Wallet struct {
	ID       string
	Currency string
	Balance  Money
}

// Debit deducts the provided amount from the wallet balance.
func (w *Wallet) Debit(amount Money) error {
	if w.Currency != amount.Currency {
		return ErrCurrencyMismatchWallet
	}
	if w.Balance.Amount.LessThan(amount.Amount) {
		return ErrInsufficientFunds
	}
	newBalance := w.Balance.Amount.Sub(amount.Amount)
	w.Balance.Amount = newBalance
	return nil
}

// Credit adds the provided amount to the wallet balance.
func (w *Wallet) Credit(amount Money) error {
	if w.Currency != amount.Currency {
		return ErrCurrencyMismatchWallet
	}
	w.Balance.Amount = w.Balance.Amount.Add(amount.Amount)
	return nil
}
